### ISSUE TITLE
[Off-Topic] Use sympy to provide symbolic computation in R

### ISSUE THREAD

**Issue by Marlin-Na** [rstudio/reticulate#29]
Firstly I want to really thank Allaire and RStudio people for all their significant works, including this remarkable package. It is also good to know that reticulate is already on CRAN.

An idea in my mind is to bring symbolic computation capacity to R, possibly by wrapping the [sympy](http://www.sympy.org/en/index.html) python library. We know that R is not designed for symbolic computation, however it does not limit the attempt to bring such capacity. One example that I previously looking at is the [SymPy](https://github.com/JuliaPy/SymPy.jl) package for julia, but I think it is possible to provide a more seamlessly integration of sympy to R, with reticulate. And I think it is even possible that the R's wrapper can be more convenient than the python package itself -- due to the lazy evaluation nature of R, we may use some syntax magic to define a symbolic expression without declaring variables, etc..

In fact, reticulate already provides seamlessly interface to python functions, we can further substitute R's binary operators and math functions using group generics in order to work with sympy variables. We may also provide ways to convert a symbolic expression to a R function (as the julia package SymPy does) and other convenient functions to work with symbolic computation.

I have done some experiment with the idea (see https://github.com/Marlin-Na/symbolicR), due to the limited time and also my limited experience, I have not managed to provide a relative complete prototype so far.

But my question is:

How do you think about the idea that bring sympy to R with reticulate? Or did you have such plan when you created the reticulate package?

Thanks!

**Comment by jjallaire** [comment]
My thoughts on that subject in general are here: https://github.com/rstudio/reticulate/issues/14

TL;DR is that we should create bridges only when we absolutely have to (i.e. there is no native code equivalent to bind to) mostly because we don't want to burden users with having to maintain two runtimes and two parallel sets of packages dependencies. Your case may indeed fall within this category (I'm not familiar with what other symbolic computing libraries are out there to build on top of).

**Comment by eddelbuettel** [comment]
Just for completeness, there is an older R package on CRAN  which bring SymPy to R: [rSymPy](https://cran.r-project.org/web/packages/rSymPy/index.html)

It is based on the R-to-Jython Java/Python interface, which is probably no longer all that alive.

**Comment by jjallaire** [comment]
Going to close this b/c it's not an issue per-se. Naturally feel free to continue the discussion here as required!

**Comment by Marlin-Na** [comment]
Thanks very much, @jjallaire and also thank @eddelbuettel for the info!

I agree with the point and that reminds me of [symengine](https://github.com/symengine/symengine), which is a symbolic manipulation library written in C++. It is pretty much a faster swappable core of sympy and there are wrappers written in other languages (julia, ruby and haskell). I believe it may be more suitable to be ported in R.

But unfortunately I am not familiar with C++......
Hopefully I can learn it in the future or look for help.

Thanks again;

**Comment by eddelbuettel** [comment]
Dang.  That would have made a nice Google Summer of Code project.  Maybe next year...

**Comment by Marlin-Na** [comment]
@eddelbuettel Perhaps we can first add it on the wiki to see if there are any passionate student..  Let me count, there are still one more week for the deadline... :smile:

**Comment by eddelbuettel** [comment]
I'm already listed twice as a maintainer but if you have spare cycles for this by all means give it a shot.  Happy to help with occasional Rcpp question but cannot co-mentor.

**Comment by Marlin-Na** [comment]
@eddelbuettel Anyway, i'm happy to discuss and also thanks for your work on Rcpp. I have tried to communicate with the symengine people to see if they are interested in the idea.

Best,
